Operators

icecream17 opened the pull request:
- <s>Previously blocked by https://github.com/tree-sitter/tree-sitter-javascript/pull/152</s> (and other updating prs)
- <s>Previously blocked by #692</s>
    - Blocked until https://github.com/atom/atom/issues/22129 is fixed
    - Well, it works in legacy tree-mate, but not in tree-sitter.
    - Help is very appreciated

# Description of the Change
Edit: update

Adds optional chaining `a?.b` `a?.[5]` `a?.()`
Adds `**=`
Adds logical assignment `&&= ||= ??=`
Add `shift` class to `>>= >>>= <<=`
Add `**`

Ugh unfinished edit this later

### Alternate Designs
<s>I don't see how to make the ?. and . distinct without making everything that uses it seperate:

It's much bigger than this but the idea is the same:

Original:
```"var.prop" -> "variable" "meta.delimiter.period" "property"```

Two way I thought of to support ?.
```"var(?. or .)prop" -> "var" "meta.delimiter" "property"```

or
```
"var.prop" -> "var" "meta.delimiter.period" "property"
"var?.prop" -> "var" "meta.delimiter.optional" "property"
```

I still added ?. as "meta.delimiter.optional" in the tree-sitter, but in the non-tree-sitter, they're the same.</s>
Edit: They're separate selectors, there's [a later update below](https://github.com/atom/language-javascript/pull/686#issuecomment-770310223).

I've added tests for the ?. operator, and included ?? in the logic operator test. (Works!!!!!)

### Benefits
Fixes #680
Fixes #640
Fixes #715 (duplicate)

### Possible Drawbacks
A lot of comments
<s>There might be a way to make them separate without adding a bunch of redundant code, but I haven't figured it out yet.</s> I actually did figure it out, again there's an update below

### Applicable Issues
#640 #680

icecream17 committed:
Update jsdoc-spec.coffee

icecream17 committed:
Update javascript-spec.coffee

icecream17 committed:
Update tree-sitter-javascript.cson

icecream17 committed:
Update javascript.cson

icecream17 committed:
More accurate comment

What a silly commit

icecream17 commented:
Only seems to fix ?? right now, so the current code doesn't do what I thought (of course...)
![image](https://user-images.githubusercontent.com/58114641/101549666-aeb43500-3973-11eb-8b8b-620e1aee5a0e.png)

It thinks the ? is ternary, and the period to be a.. regular period.

icecream17 committed:
Merge pull request #2 from atom/master

Update 2 merge good

icecream17 commented:
Oh the latest issue made me realize I didn't even put in
hmm?.[2]
and
hmm.jump?.()

I think I should reset the files to atom:master to retry fixing this

utkarshgupta137 commented:
Shouldn't we also include "??=" in operators?

icecream17 committed:
Add all missing operators

"delete" is not listed as an operator like typeof... not changing for now

icecream17 committed:
Add missing assignment operators

icecream17 committed:
Start trying to make ?. work

icecream17 committed:
test for extra operators

add "optional", "**=", and other test

icecream17 commented:
I added the missing operators but I broke something and the tests are failing

darangi commented:
Hi @icecream17 could you take a look into the failing tests?

icecream17 committed:
what am i doing

icecream17 commented:
Current status
exampleFunc?.() not supported
exampleArr[7]?.() not supported
a?.b should be support if it works
a?.b() should also be supported if it works

> 
> 
> Shouldn't we also include "??=" in operators?

Oh yeah. That's added
As for the errors, I try to figure them out tomorrow - I don't know what's wrong, except that regular unrelated code isnt parsing so there's some catastrophic error somewhere

icecream17 committed:
add end quote

icecream17 committed:
add Another quote!

ok no syntax stuff now

icecream17 committed:
add .period / .optional [skip ci]

See the previous result

icecream17 committed:
use #includes instead

icecream17 commented:
359 fails, now down to 66, though https://github.com/atom/language-javascript/pull/686/commits/6f32f5ad84fe1ca2034c104a27e34c39a4ec7c3a has only 13 errors, so maybe I should just use that instead....

maybe I should provide less updates?

icecream17 committed:
only fixes 1 error

icecream17 committed:
fix caputres

icecream17 committed:
update name [skip ci]

icecream17 committed:
simplify

icecream17 committed:
oops wrong one

icecream17 committed:
include doesn't give a name....

very verbose code!

icecream17 committed:
use outer strategies instead?

fixes something at line 782...

icecream17 committed:
undo last 2 commits, but keep line 740 fix

icecream17 committed:
nvm / found different fix

742 was on purpose, meanwhile some regex weren't correct because of my sloppy editing

icecream17 committed:
add ?.[8]

icecream17 committed:
add ?.()

icecream17 committed:
possibly fix one of the 3 errors

icecream17 committed:
oh *method*

simplify function_calls also
previous commit was wrong

icecream17 committed:
fix 1 spec error

icecream17 committed:
add exponentation [skip ci]

icecream17 committed:
add exponentation

icecream17 committed:
Add meta.method-call and some extra tests

wow

icecream17 committed:
avoid 0: '*', 1: '*'

icecream17 committed:
spec error

icecream17 committed:
support a?.[8]

icecream17 committed:
support [array]?.()

icecream17 commented:
It passed!!!!!!

Summary of changes:
![image](https://user-images.githubusercontent.com/58114641/106372048-5ecd6b00-6331-11eb-9960-fc893e60a1e7.png)

Non tree-sitter scope changes:
```
optional matches "?."

meta.delimiter.method.js -->
   meta.delimiter.method.period.js
   meta.delimiter.method.optional.js
meta.delimiter.property.js -->
   meta.delimiter.property.period.js
   meta.delimiter.property.optional.js

keyword.operator.assignment.compound.logical.js added
   used by &&= ||= ??=
```

icecream17 committed:
remove todo

icecream17 commented:
Oh wait, it only works in the Non-Tree Sitter Parser

icecream17 committed:
try some silly precedence changes [skip-ci skip ci]

icecream17 committed:
maybe this will work

icecream17 committed:
Undo last 2 commits

It's possible that the eslint package messed things up....

icecream17 committed:
Try the ? again

sadick254 commented on grammars/javascript.cson:
```diff
@@ -1,3 +1,6 @@
+# Note: Because of the optional chaining operator (?.), "object.property" in the comments usually means ("object.property" or "object?.property")
+# Even though sometimes it doesn't seem like using the chaining operator would make any sense, it's still supported. example: Math?.PI
+# Negative lookbehinds for "." Shouldddd cover "?."
```
```suggestion
# Negative look-behinds for "." Should cover "?."
```

icecream17 committed:
That was a weird comment

Thanks

Co-authored-by: Sadick <sadickjunior@gmail.com>

icecream17 committed:
Try reordering the operators instead. Add void.

To check later (I don't have atom on this machine)

The colon, which is part of the ternary operator, may also be parsed at part of a:
label - https://github.com/tree-sitter/tree-sitter-javascript/blob/master/src/grammar.json#L1304 
switch case - https://github.com/tree-sitter/tree-sitter-javascript/blob/master/src/grammar.json#L1359
switch default - https://github.com/tree-sitter/tree-sitter-javascript/blob/master/src/grammar.json#L1380

Generators, e.g. function*
https://github.com/tree-sitter/tree-sitter-javascript/blob/master/src/grammar.json#L2676

A JSX fragment, ```<></>```, is also in the grammar, and will obviously fail.

Resources I found
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Operator_Precedence
https://github.com/tree-sitter/tree-sitter-javascript/blob/master/src/grammar.json with ctrl+f "type": "STRING"

icecream17 commented:
There's some precedence thing with the operator parsing. I'm not sure how it works.

I found the actual parser: https://github.com/tree-sitter/tree-sitter-javascript
And the grammar file: https://github.com/tree-sitter/tree-sitter-javascript/blob/master/src/grammar.json
_______
<sup>Note to self: https://github.com/atom/language-javascript/blob/master/grammars/tree-sitter-javascript.cson just maps parsed nodes to names.<sup>

icecream17 committed:
try doing the reverse order, fix ?:

The colon is used for a lot of different things. I'm using what the non-tree sitter labels them. Although there's probably better names.

icecream17 committed:
Some operators don't exist

icecream17 committed:
Nevermind, they do exist

The atom console is outdated

icecream17 committed:
order doesn't matter, visual appeal [skip ci]

This is a visual commit.
Really it's about precedence, how specific the selectors are. Since they're the same it should work no matter what.

The reason ?. isn't detected is because of the parser itself.

See https://github.com/jcs-PR/tree-sitter-javascript/pull/1/files and https://github.com/tree-sitter/tree-sitter-javascript/pull/150

icecream17 committed:
slight visual change [skip ci]

icecream17 commented:
![image](https://user-images.githubusercontent.com/58114641/107729200-d4342680-6cb5-11eb-8d31-1f7e15f8762d.png)

The tree-sitter parser atom uses can't actually parse some operators yet. Will update when it can.

icecream17 committed:
Merge pull request #3 from atom/master

update

icecream17 committed:
add rest and bitwise shift

icecream17 committed:
correct get, add scopes, consistency [skip ci]

icecream17 committed:
add rest matcher

icecream17 committed:
undo rest, try to fix infinite loop

icecream17 committed:
...i figured out the infinite loop

Empty string "" times infinity is why
Undo the undo

icecream17 committed:
update spec to match

icecream17 committed:
fix operator precedence

icecream17 committed:
fix **=

icecream17 committed:
Merge pull request #4 from atom/master

Update from atom

icecream17 commented:
Update: 
See https://github.com/atom/atom/issues/22129
Because of that error it's blocked until atom is updated: https://github.com/atom/atom/pull/22130

icecream17 commented:
Ok, even though it still doesn't work in tree-sitter, it could still be merged as long as there's no bad changes.

_____
### Incompatible changes

Currently I've "added scopes" in https://github.com/atom/language-javascript/pull/686/commits/743fcd5b42fe9efe1f91c61ea7c9aee0c089f946 but https://github.com/atom/language-javascript/pull/690/files takes care of that.
